utility/src/main/java/org/oppia/android/util/logging/performancemetrics/PerformanceMetricsAssessorImpl.kt

```diff
@@ -70,4 +77,49 @@ class PerformanceMetricsAssessorImpl @Inject constructor(
       else -> HIGH_MEMORY_TIER
     }
   }
+
+  override fun computeCpuSnapshotAtCurrentTime(): CpuSnapshot {
+    return CpuSnapshot(
+      appTimeMillis = oppiaClock.getCurrentTimeMs(),
+      cpuTimeMillis = Process.getElapsedCpuTime(),
+      numberOfOnlineCores = getNumberOfOnlineCores()
+    )
+  }
+
+  override fun getRelativeCpuUsage(
+    firstCpuSnapshot: CpuSnapshot,
+    secondCpuSnapshot: CpuSnapshot
+  ): Double? {
+    val deltaCpuTimeMs = if (secondCpuSnapshot.cpuTimeMillis >= firstCpuSnapshot.cpuTimeMillis) {
+      secondCpuSnapshot.cpuTimeMillis - firstCpuSnapshot.cpuTimeMillis
+    } else { return null }
+    val deltaProcessTimeMs = if (secondCpuSnapshot.appTimeMillis > firstCpuSnapshot.appTimeMillis) {
+      secondCpuSnapshot.appTimeMillis - firstCpuSnapshot.appTimeMillis
+    } else { return null }
+    val numberOfCores =
+      if (secondCpuSnapshot.numberOfOnlineCores >= 1 && firstCpuSnapshot.numberOfOnlineCores >= 1) {
+        (secondCpuSnapshot.numberOfOnlineCores + firstCpuSnapshot.numberOfOnlineCores) / 2.0
+      } else { return null }
+
+    return when (val relativeCpuUsage = deltaCpuTimeMs / (deltaProcessTimeMs * numberOfCores)) {
+      in 0.0..1.0 -> relativeCpuUsage
+      else -> { null }
+    }
+  }
```

Comment:
```suggestion
  override fun getRelativeCpuUsage(
    firstCpuSnapshot: CpuSnapshot,
    secondCpuSnapshot: CpuSnapshot
  ): Double? {
    if (firstCpuSnapshot.isNewerThan(secondCpuSnapshot)) return null
    val deltaCpuTimeMs = secondCpuSnapshot.cpuTimeMillis - firstCpuSnapshot.cpuTimeMillis
    val deltaProcessTimeMs = secondCpuSnapshot.appTimeMillis - firstCpuSnapshot.appTimeMillis
    val numberOfCores =
      (secondCpuSnapshot.numberOfOnlineCores + firstCpuSnapshot.numberOfOnlineCores) / 2.0

    return when (val relativeCpuUsage = deltaCpuTimeMs / (deltaProcessTimeMs * numberOfCores)) {
      in 0.0..1.0 -> relativeCpuUsage
      else -> null
    }
  }
  
  ...
  
  // To go within CpuSnapshot.
  fun isNewerThan(other: CpuSnapshot) =
    cpuTimeMillis > other.cpuTimeMillis || appTimeMillis > other.appTimeMillis
```

Some possible simplification. This is also assuming that we coerce the number of cores to always be 1 or larger (which I think is reasonable since any value under 1 is simply invalid/impossible to occur).

Reply:
Nice suggestion. However, I've extended this a bit to also have a `doesNotHaveValidNumberOfOnlineCores` for the current CPU snapshot as well. I think any coercion is wrong here because if the system is returning less than 1 CPU cores then it's simply also an indicator that the entire cpu reading may be false at that time -- in short, I don't think we should "fix" any value on our own and rather not log in case of invalid input.